drivers/mpsl/clock_control/clock_control_mpsl.c

```diff
@@ -207,30 +275,80 @@ static enum clock_control_status clock_get_status(struct device *dev,
 	}
 }
 
+static int clock_async_start(struct device *dev, clock_control_subsys_t subsys,
+			     struct clock_control_async_data *data)
+{
+	struct mpsl_clock_control_data *mpsl_control_data = dev->driver_data;
+	enum clock_control_status clock_status = clock_get_status(dev, subsys);
+	int key;
+
+	switch (clock_status) {
+	case CLOCK_CONTROL_STATUS_ON:
+		/* LFCLK->Clock is started by default and not able to stop it.
+		 * Therefore, LFCLK->Clock is always handled in case STATUS_ON.
+		 */
+		if ((data != NULL) && (data->cb != NULL)) {
+			data->cb(dev, data->user_data);
+		}
+
+		if (subsys == CLOCK_CONTROL_NRF_SUBSYS_HF) {
+			key = irq_lock();
+			++mpsl_control_data->hfclk_start_count;
+			irq_unlock(key);
+		}
+
+		return 0;
+	case CLOCK_CONTROL_STATUS_STARTING:
+	case CLOCK_CONTROL_STATUS_OFF:
+		if ((data != NULL) && (data->cb != NULL)) {
```

Comment:
if clock interrupt comes here and clock state changes to on then adding to list will not result in user notification. I guess you need to lock for status check and adding to the list

Reply:
I see. Is there any side-effect if it is looping all the user callbacks while irq is locked?

Reply:
well, hard to say. it is definitely dangerous.

Reply:
Agree. However, it is also important to ensure that the clock keeps on when running in the callback. As stated in clock_control.h, the callback may be called from the context of clock interrupt. Let's assume that the user won't do sleep or polling status in the callback.